File: src/main/java/mnm/mods/tabbychat/client/core/GuiChatTC.java
package mnm.mods.tabbychat.client.core;

import mnm.mods.tabbychat.client.AbstractChannel;
import mnm.mods.tabbychat.client.TabbyChatClient;
import mnm.mods.tabbychat.client.gui.ChatBox;
import mnm.mods.tabbychat.client.gui.component.GuiText;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiChat;
import net.minecraft.client.gui.IGuiEventListener;
import net.minecraftforge.client.event.GuiScreenEvent;
import net.minecraftforge.eventbus.api.EventPriority;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.TickEvent;
import org.lwjgl.glfw.GLFW;

import java.util.List;

public class GuiChatTC {

    private final ChatBox chat;

    public GuiChatTC(ChatBox chat) {
        this.chat = chat;
    }

    @SuppressWarnings("unchecked")
    @SubscribeEvent
    public void initGui(GuiScreenEvent.InitGuiEvent.Post event) {
        if (event.getGui() instanceof GuiChat) {
            GuiChat guichat = (GuiChat) event.getGui();
            AbstractChannel chan = chat.getActiveChannel();
            if (guichat.defaultInputFieldText.isEmpty()
                    && !chan.isPrefixHidden()
                    && !chan.getPrefix().isEmpty()) {
                guichat.defaultInputFieldText = chan.getPrefix() + " ";
            }
            GuiText text = chat.getChatInput().getTextField();
            guichat.inputField = text.getTextField();
            text.setValue(guichat.defaultInputFieldText);

            text.getTextField().setTextFormatter(guichat::formatMessage);
            text.getTextField().setTextAcceptHandler(guichat::acceptMessage);

            List<IGuiEventListener> children = (List<IGuiEventListener>) guichat.getChildren();
            children.set(0, chat);
        }
    }

    @SubscribeEvent
    public void onTick(TickEvent.ClientTickEvent event) {
        if (Minecraft.getInstance().currentScreen instanceof GuiChat && event.phase == TickEvent.Phase.END) {
            chat.tick();
        }
    }

    @SubscribeEvent(priority = EventPriority.LOWEST)
    public void onRenderChat(GuiScreenEvent.DrawScreenEvent.Pre event) {
        if (event.getGui() instanceof GuiChat) {
            event.setCanceled(true);
            this.chat.render(event.getMouseX(), event.getMouseY(), event.getRenderPartialTicks());
        }
    }

    @SubscribeEvent
    public void onKeyPressed(GuiScreenEvent.KeyboardKeyPressedEvent.Pre event) {
        if (event.getGui() instanceof GuiChat) {
            if (keyPressed((GuiChat) event.getGui(), event.getKeyCode())
                    || this.chat.keyPressed(event.getKeyCode(), event.getScanCode(), event.getModifiers())) {
                event.setCanceled(true);
            }
        }
    }

    @SubscribeEvent
    public void onKeyReleased(GuiScreenEvent.KeyboardKeyReleasedEvent.Pre event) {
        if (event.getGui() instanceof GuiChat) {
            if (this.chat.keyPressed(event.getKeyCode(), event.getScanCode(), event.getModifiers())) {
                event.setCanceled(true);
            }
        }
    }

    @SubscribeEvent
    public void onCharTyped(GuiScreenEvent.KeyboardCharTypedEvent.Pre event) {
        if (event.getGui() instanceof GuiChat) {
            if (this.chat.charTyped(event.getCodePoint(), event.getModifiers())) {
                event.setCanceled(true);
            }
        }
    }

    @SubscribeEvent
    public void onMouseClicked(GuiScreenEvent.MouseClickedEvent.Pre event) {
        if (event.getGui() instanceof GuiChat) {
            if (this.chat.mouseClicked(event.getMouseX(), event.getMouseY(), event.getButton())) {
                event.setCanceled(true);
            }
        }
    }
    @SubscribeEvent
    public void onMouseReleased(GuiScreenEvent.MouseReleasedEvent.Pre event) {
        if (event.getGui() instanceof GuiChat) {
            if (this.chat.mouseReleased(event.getMouseX(), event.getMouseY(), event.getButton())) {
                event.setCanceled(true);
            }
        }
    }

    @SubscribeEvent
    public void onMouseDragged(GuiScreenEvent.MouseDragEvent.Pre event) {
        if (event.getGui() instanceof GuiChat) {
            if (this.chat.mouseDragged(event.getMouseX(), event.getMouseY(), event.getMouseButton(), event.getDragX(), event.getDragY())) {
                event.setCanceled(true);
            }
        }
    }


    @SubscribeEvent
    public void onMouseScrolled(GuiScreenEvent.MouseScrollEvent.Pre event) {
        if (event.getGui() instanceof GuiChat) {
            if (this.chat.mouseScrolled(event.getScrollDelta())) {
                event.setCanceled(true);
            }
        }
    }

    private boolean keyPressed(GuiChat guichat, int key) {
        if (key == GLFW.GLFW_KEY_ENTER || key == GLFW.GLFW_KEY_KP_ENTER) {
            Minecraft.getInstance().ingameGUI.getChatGUI().resetScroll();
            GuiText text = this.chat.getChatInput().getTextField();
            guichat.sendChatMessage(text.getValue());
            text.setValue(guichat.defaultInputFieldText);

            if (!TabbyChatClient.getInstance().getSettings().advanced.keepChatOpen.get()) {
                Minecraft.getInstance().displayGuiScreen(null);
            }
            return true;
        }
        return false;
    }
}

File: src/main/kotlin/mnm/mods/tabbychat/client/core/GuiChatTC.kt
package mnm.mods.tabbychat.client.core

import mnm.mods.tabbychat.client.TabbyChatClient
import mnm.mods.tabbychat.client.gui.ChatBox
import mnm.mods.tabbychat.client.gui.suggestionHelper
import mnm.mods.tabbychat.util.mc
import net.minecraft.client.gui.IGuiEventListener
import net.minecraft.client.gui.screen.ChatScreen
import net.minecraftforge.client.event.GuiScreenEvent
import net.minecraftforge.event.TickEvent
import net.minecraftforge.eventbus.api.EventPriority
import net.minecraftforge.eventbus.api.SubscribeEvent
import org.lwjgl.glfw.GLFW

object GuiChatTC {

    @SubscribeEvent
    fun initGui(event: GuiScreenEvent.InitGuiEvent.Post) {
        if (event.gui is ChatScreen) {
            val guichat = event.gui as ChatScreen
            val chan = ChatBox.activeChannel
            if (guichat.defaultInputFieldText.isEmpty()
                    && !chan.isPrefixHidden
                    && chan.prefix.isNotEmpty()) {
                guichat.defaultInputFieldText = chan.prefix + " "
            }
            val text = ChatBox.chatInput.textField
            guichat.inputField = text.delegate
            text.value = guichat.defaultInputFieldText

            ChatBox.chatInput.textFormatter = guichat.suggestionHelper::func_228122_a_
            text.delegate.setResponder { guichat.func_212997_a(it) }

            val children = guichat.children() as MutableList<IGuiEventListener>
            children[0] = ChatBox
        }
    }

    @SubscribeEvent
    fun onTick(event: TickEvent.ClientTickEvent) {
        if (mc.currentScreen is ChatScreen && event.phase == TickEvent.Phase.END) {
            ChatBox.tick()
        }
    }

    @SubscribeEvent(priority = EventPriority.LOWEST)
    fun onRenderChat(event: GuiScreenEvent.DrawScreenEvent.Pre) {
        if (event.gui is ChatScreen) {
            event.isCanceled = true
            ChatBox.update(event.gui as ChatScreen)
            ChatBox.render(event.mouseX, event.mouseY, event.renderPartialTicks)
        }
    }

    @SubscribeEvent
    fun onKeyPressed(event: GuiScreenEvent.KeyboardKeyPressedEvent.Pre) {
        if (event.gui is ChatScreen) {
            event.isCanceled = keyPressed(event.gui as ChatScreen, event.keyCode) || ChatBox.keyPressed(event.keyCode, event.scanCode, event.modifiers)
        }
    }

    @SubscribeEvent
    fun onKeyReleased(event: GuiScreenEvent.KeyboardKeyReleasedEvent.Pre) {
        if (event.gui is ChatScreen) {
            event.isCanceled = ChatBox.keyReleased(event.keyCode, event.scanCode, event.modifiers)
        }
    }

    @SubscribeEvent
    fun onCharTyped(event: GuiScreenEvent.KeyboardCharTypedEvent.Pre) {
        if (event.gui is ChatScreen) {
            event.isCanceled = ChatBox.charTyped(event.codePoint, event.modifiers)
        }
    }

    @SubscribeEvent
    fun onMouseClicked(event: GuiScreenEvent.MouseClickedEvent.Pre) {
        if (event.gui is ChatScreen) {
            event.isCanceled = ChatBox.mouseClicked(event.mouseX, event.mouseY, event.button)
        }
    }

    @SubscribeEvent
    fun onMouseReleased(event: GuiScreenEvent.MouseReleasedEvent.Pre) {
        if (event.gui is ChatScreen) {
            event.isCanceled = ChatBox.mouseReleased(event.mouseX, event.mouseY, event.button)
        }
    }

    @SubscribeEvent
    fun onMouseDragged(event: GuiScreenEvent.MouseDragEvent.Pre) {
        if (event.gui is ChatScreen) {
            event.isCanceled = ChatBox.mouseDragged(event.mouseX, event.mouseY, event.mouseButton, event.dragX, event.dragY)
        }
    }

    @SubscribeEvent
    fun onMouseScrolled(event: GuiScreenEvent.MouseScrollEvent.Pre) {
        if (event.gui is ChatScreen) {
            event.isCanceled = ChatBox.mouseScrolled(event.mouseX, event.mouseY, event.scrollDelta)
        }
    }

    private fun keyPressed(guichat: ChatScreen, key: Int): Boolean {
        if (key == GLFW.GLFW_KEY_ENTER || key == GLFW.GLFW_KEY_KP_ENTER) {
            mc.ingameGUI.chatGUI.resetScroll()
            val text = ChatBox.chatInput.textField
            guichat.sendMessage(text.value)
            text.value = guichat.defaultInputFieldText

            if (!TabbyChatClient.settings.advanced.keepChatOpen.value) {
                mc.displayGuiScreen(null)
            }
            return true
        }
        return false
    }
}